File: src/commerce/fraud-detection-service/fraud/store.py
"""asyncpg-backed persistence for fraud check results."""

from __future__ import annotations

import json
from typing import Optional

import asyncpg

from fraud.models import FraudCheckResult

_MIGRATION_SQL = """
CREATE TABLE IF NOT EXISTS fraud_results (
    order_id        TEXT PRIMARY KEY,
    risk_score      INTEGER NOT NULL,
    risk_level      TEXT NOT NULL,
    decision        TEXT NOT NULL,
    signals         JSONB NOT NULL DEFAULT '[]',
    checked_at      TIMESTAMPTZ NOT NULL
);
CREATE INDEX IF NOT EXISTS idx_fraud_results_decision ON fraud_results (decision);
CREATE INDEX IF NOT EXISTS idx_fraud_results_checked_at ON fraud_results (checked_at DESC);
"""


class FraudStore:
    def __init__(self, pool: asyncpg.Pool) -> None:
        self._pool = pool

    @classmethod
    async def create(cls, database_url: str) -> "FraudStore":
        pool = await asyncpg.create_pool(database_url, min_size=2, max_size=10)
        store = cls(pool)
        await store._migrate()
        return store

    async def _migrate(self) -> None:
        async with self._pool.acquire() as conn:
            await conn.execute(_MIGRATION_SQL)

    async def close(self) -> None:
        await self._pool.close()

    async def save_result(self, result: FraudCheckResult) -> None:
        sql = """
            INSERT INTO fraud_results
                (order_id, risk_score, risk_level, decision, signals, checked_at)
            VALUES ($1, $2, $3, $4, $5::jsonb, $6)
            ON CONFLICT (order_id) DO UPDATE SET
                risk_score  = EXCLUDED.risk_score,
                risk_level  = EXCLUDED.risk_level,
                decision    = EXCLUDED.decision,
                signals     = EXCLUDED.signals,
                checked_at  = EXCLUDED.checked_at
        """
        async with self._pool.acquire() as conn:
            await conn.execute(
                sql,
                result.order_id,
                result.risk_score,
                result.risk_level,
                result.decision,
                json.dumps(result.signals),
                result.checked_at,
            )

    async def get_result(self, order_id: str) -> Optional[FraudCheckResult]:
        sql = """
            SELECT order_id, risk_score, risk_level, decision, signals, checked_at
            FROM fraud_results
            WHERE order_id = $1
        """
        async with self._pool.acquire() as conn:
            row = await conn.fetchrow(sql, order_id)
        if row is None:
            return None
        return FraudCheckResult(
            order_id=row["order_id"],
            risk_score=row["risk_score"],
            risk_level=row["risk_level"],
            decision=row["decision"],
            signals=json.loads(row["signals"]),
            checked_at=row["checked_at"],
        )

    async def list_flagged(self, limit: int = 50) -> list[FraudCheckResult]:
        sql = """
            SELECT order_id, risk_score, risk_level, decision, signals, checked_at
            FROM fraud_results
            WHERE decision IN ('review', 'decline')
            ORDER BY checked_at DESC
            LIMIT $1
        """
        async with self._pool.acquire() as conn:
            rows = await conn.fetch(sql, limit)
        return [
            FraudCheckResult(
                order_id=r["order_id"],
                risk_score=r["risk_score"],
                risk_level=r["risk_level"],
                decision=r["decision"],
                signals=json.loads(r["signals"]),
                checked_at=r["checked_at"],
            )
            for r in rows
        ]

File: src/commerce/ab-testing-service/store/store.go
package store

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"time"

	_ "github.com/lib/pq"

	"github.com/shopos/ab-testing-service/domain"
)

const migrationSQL = `
CREATE TABLE IF NOT EXISTS experiments (
    id               TEXT PRIMARY KEY,
    name             TEXT NOT NULL,
    description      TEXT NOT NULL DEFAULT '',
    variants         JSONB NOT NULL DEFAULT '[]',
    active           BOOLEAN NOT NULL DEFAULT TRUE,
    traffic_percent  INTEGER NOT NULL DEFAULT 100,
    created_at       TIMESTAMPTZ NOT NULL,
    updated_at       TIMESTAMPTZ NOT NULL
);

CREATE TABLE IF NOT EXISTS assignments (
    experiment_id  TEXT NOT NULL,
    user_id        TEXT NOT NULL,
    variant        TEXT NOT NULL,
    assigned_at    TIMESTAMPTZ NOT NULL,
    PRIMARY KEY (experiment_id, user_id)
);

CREATE TABLE IF NOT EXISTS conversions (
    id             BIGSERIAL PRIMARY KEY,
    experiment_id  TEXT NOT NULL,
    user_id        TEXT NOT NULL,
    metric         TEXT NOT NULL,
    value          DOUBLE PRECISION NOT NULL DEFAULT 0,
    recorded_at    TIMESTAMPTZ NOT NULL
);
CREATE INDEX IF NOT EXISTS idx_conversions_experiment ON conversions (experiment_id);
`

// Store handles all persistence for the ab-testing-service.
type Store struct {
	db *sql.DB
}

// New opens a Postgres connection, runs migrations, and returns a Store.
func New(databaseURL string) (*Store, error) {
	db, err := sql.Open("postgres", databaseURL)
	if err != nil {
		return nil, fmt.Errorf("open db: %w", err)
	}
	db.SetMaxOpenConns(10)
	db.SetMaxIdleConns(5)
	db.SetConnMaxLifetime(time.Hour)

	db.Exec(migrationSQL) //nolint:errcheck — best-effort at startup; fails gracefully when DB is unavailable
	return &Store{db: db}, nil
}

// Close shuts down the database pool.
func (s *Store) Close() error { return s.db.Close() }

// SaveExperiment inserts or updates an experiment.
func (s *Store) SaveExperiment(ctx context.Context, e *domain.Experiment) error {
	variantsJSON, err := json.Marshal(e.Variants)
	if err != nil {
		return err
	}
	_, err = s.db.ExecContext(ctx, `
		INSERT INTO experiments (id, name, description, variants, active, traffic_percent, created_at, updated_at)
		VALUES ($1,$2,$3,$4::jsonb,$5,$6,$7,$8)
		ON CONFLICT (id) DO UPDATE SET
		    name=EXCLUDED.name, description=EXCLUDED.description,
		    variants=EXCLUDED.variants, active=EXCLUDED.active,
		    traffic_percent=EXCLUDED.traffic_percent, updated_at=EXCLUDED.updated_at`,
		e.ID, e.Name, e.Description, string(variantsJSON),
		e.Active, e.TrafficPercent, e.CreatedAt, e.UpdatedAt,
	)
	return err
}

// GetExperiment retrieves an experiment by ID.
func (s *Store) GetExperiment(ctx context.Context, id string) (*domain.Experiment, error) {
	row := s.db.QueryRowContext(ctx, `
		SELECT id, name, description, variants, active, traffic_percent, created_at, updated_at
		FROM experiments WHERE id=$1`, id)
	return scanExperiment(row)
}

// ListExperiments returns all experiments ordered by created_at desc.
func (s *Store) ListExperiments(ctx context.Context) ([]*domain.Experiment, error) {
	rows, err := s.db.QueryContext(ctx, `
		SELECT id, name, description, variants, active, traffic_percent, created_at, updated_at
		FROM experiments ORDER BY created_at DESC`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var result []*domain.Experiment
	for rows.Next() {
		e, err := scanExperiment(rows)
		if err != nil {
			return nil, err
		}
		result = append(result, e)
	}
	return result, rows.Err()
}

type scanner interface {
	Scan(dest ...any) error
}

func scanExperiment(s scanner) (*domain.Experiment, error) {
	var e domain.Experiment
	var variantsJSON string
	err := s.Scan(
		&e.ID, &e.Name, &e.Description, &variantsJSON,
		&e.Active, &e.TrafficPercent, &e.CreatedAt, &e.UpdatedAt,
	)
	if err == sql.ErrNoRows {
		return nil, domain.ErrNotFound
	}
	if err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(variantsJSON), &e.Variants); err != nil {
		return nil, err
	}
	return &e, nil
}

// SaveAssignment inserts or updates a user-to-variant assignment.
func (s *Store) SaveAssignment(ctx context.Context, a *domain.Assignment) error {
	_, err := s.db.ExecContext(ctx, `
		INSERT INTO assignments (experiment_id, user_id, variant, assigned_at)
		VALUES ($1,$2,$3,$4)
		ON CONFLICT (experiment_id, user_id) DO NOTHING`,
		a.ExperimentID, a.UserID, a.Variant, a.AssignedAt,
	)
	return err
}

// GetAssignment retrieves an existing assignment.
func (s *Store) GetAssignment(ctx context.Context, experimentID, userID string) (*domain.Assignment, error) {
	var a domain.Assignment
	err := s.db.QueryRowContext(ctx, `
		SELECT experiment_id, user_id, variant, assigned_at
		FROM assignments WHERE experiment_id=$1 AND user_id=$2`,
		experimentID, userID,
	).Scan(&a.ExperimentID, &a.UserID, &a.Variant, &a.AssignedAt)
	if err == sql.ErrNoRows {
		return nil, domain.ErrNotFound
	}
	return &a, err
}

// RecordConversion saves a conversion event.
func (s *Store) RecordConversion(ctx context.Context, c *domain.Conversion) error {
	_, err := s.db.ExecContext(ctx, `
		INSERT INTO conversions (experiment_id, user_id, metric, value, recorded_at)
		VALUES ($1,$2,$3,$4,$5)`,
		c.ExperimentID, c.UserID, c.Metric, c.Value, c.RecordedAt,
	)
	return err
}
